ssl/record/rec_layer_s3.c

```diff
@@ -1534,50 +1535,53 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
                          SSL_R_TOO_MANY_WARN_ALERTS);
                 return -1;
             }
+        }
+
+        if (alert_descr == SSL_AD_CLOSE_NOTIFY
+                && (SSL_IS_TLS13(s) || alert_level == SSL3_AL_WARNING)) {
+            s->shutdown |= SSL_RECEIVED_SHUTDOWN;
+            return 0;
+        }
+
+        /*
+         * This is a warning but we receive it if we requested
+         * renegotiation and the peer denied it. Terminate with a fatal
+         * alert because if application tried to renegotiate it
+         * presumably had a good reason and expects it to succeed. In
+         * future we might have a renegotiation where we don't care if
+         * the peer refused it where we carry on.
+         */
+        if (alert_descr == SSL_AD_NO_RENEGOTIATION
+                && !SSL_IS_TLS13(s)
+                && alert_level == SSL3_AL_WARNING) {
+            SSLfatal(s, SSL_AD_HANDSHAKE_FAILURE, SSL_F_SSL3_READ_BYTES,
+                     SSL_R_NO_RENEGOTIATION);
+            return -1;
+        }
 
-            if (alert_descr == SSL_AD_CLOSE_NOTIFY) {
+        /*
+         * Apart from close_notify the only other warning alert in TLSv1.3
+         * is user_cancelled - which we just ignore.
+         */
+        if (!SSL_IS_TLS13(s) || alert_descr != SSL_AD_USER_CANCELLED) {
+            if (SSL_IS_TLS13(s) || alert_level == SSL3_AL_FATAL) {
```

Comment:
Not that one can directly measure the difference, but this is actually suboptimal [because SSL_IS_TLS13 is evaluated twice (though without derefencing memory) and there is a lot of branches]. Trouble is that a lot of small suboptimal choices can add up to something big (as for example in #5585). But to be frank that's not what actually what triggered this comment. It's readability. It's unnecessary hard to unravel this condition. It appears more appropriate to introduce a variable called say tls13_ad_user_cancelled, assign `SSL_IS_TLS13(s) && alert_descr == SSL_AD_USER_CANCELLED` to it, and rewrite this condition as if (alert_level == SSL3_AL_FATAL && !tls13_ad_user_cancelled)`. The variable can naturally be used even above...

Reply:
You caught me seconds away from "git push" with this :-). I'll take a look.

Reply:
Another thing that bothers me is that this is [supposed to be?] switch-like construct, yet there are no else if's. This makes you look for overlaps in conditions and question if order of if's is correct. By latter I mean that it might be appropriate to check for some conditions prior others... This is side note...